### PR TITLE
planner: fix set not null flag for outer join

### PR DESCRIPTION
**Pull request by fzhedu**
### What problem does this PR solve?

Issue Number: close https://github.com/pingcap/tidb/issues/23737, https://github.com/pingcap/tics/issues/1703  { #22941 (the previous PR https://github.com/pingcap/tidb/pull/22955 also introduce a bug) }

Problem Summary:  the previous PR https://github.com/pingcap/tidb/pull/22955  writes a wrong child id as pointed [here](https://github.com/pingcap/tidb/pull/22955/files#r604634138). Besides, `buildLogicalJoinSchema` does the same work, so I refactor it as the code of this PR.

### Check List 

Tests 

- Unit test
- Integration test
- Manual test (add detailed scripts or steps below)
- No code

### Release note 

- No release note

### CODE REVIEW

**Comment by fzhedu** [comment]
@eurekaka this issue breaks the [previous claim](https://github.com/pingcap/tidb/pull/23417#pullrequestreview-621849237) "Currently, we can only make a promise that there is no false positive indication(i.e, when a column is marked as NotNull, it must have no null values, while when a column is not marked as NotNull, it may have no null values actually)"

File: executor/index_lookup_merge_join_test.go [hunk 1]
```diff
@@ -10,7 +10,7 @@ import (
 	"github.com/pingcap/tidb/util/testkit"
 )
 
-func (s *testSuite9) TestIndexLookupMergeJoinHang(c *C) {
+func (s *testSuiteJoin2) TestIndexLookupMergeJoinHang(c *C) {
```
Comment (fzhedu) [inline executor/index_lookup_merge_join_test.go:13]: to avoid test conflict resulting invalid errors.

File: planner/core/rule_column_pruning.go [hunk 2]
```diff
@@ -303,24 +303,7 @@ func (p *LogicalJoin) extractUsedCols(parentUsedCols []*expression.Column) (left
...
-		p.schema = lChild.Schema().Clone()
-	} else if p.JoinType == LeftOuterSemiJoin || p.JoinType == AntiLeftOuterSemiJoin {
-		joinCol := p.schema.Columns[len(p.schema.Columns)-1]
-		p.schema = lChild.Schema().Clone()
-		p.schema.Append(joinCol)
-	} else {
-		p.schema = expression.MergeSchema(lChild.Schema(), rChild.Schema())
-		switch p.JoinType {
-		case LeftOuterJoin:
-			resetNotNullFlag(p.schema, p.children[1].Schema().Len(), p.schema.Len())
```
Comment (fzhedu) [inline planner/core/rule_column_pruning.go:318]: `p.children[1].Schema().Len()` should be `p.children[0].Schema().Len()`

Comment (fzhedu) [reply planner/core/rule_column_pruning.go:318]: ref https://github.com/pingcap/tidb/blob/505955ad0b6bdaf2550c1bd1ab4858834af02048/planner/core/util.go#L228

**Review by hanfei1991 — APPROVED** [review]

**Pull request merged**